src/stream/src/cache/range_lru.rs

```diff
@@ -0,0 +1,542 @@
+                    Included(x) => Some((Excluded(x), existing_range.1.clone())),
+                    Excluded(x) => Some((Included(x), existing_range.1.clone())),
+                };
+                self.existing_range = new_existing_range.clone();
+                (true, new_existing_range)
+            } else if range.1 == existing_range.1 {
+                // Evict from the top
+                let new_existing_range = match range.0.clone() {
+                    Unbounded => None,
+                    Included(x) => Some((existing_range.0.clone(), Excluded(x))),
+                    Excluded(x) => Some((existing_range.0.clone(), Included(x))),
+                };
+                self.existing_range = new_existing_range.clone();
+                (false, new_existing_range)
+            } else {
+                panic!("The evicted range {range:?} does not share a bound with the existing range {existing_range:?}");
+            };
+
+            // If from top, delete top part of the range
+            if !from_bottom {
+                // If from top remove the top marker as well.
+                match range.1.clone() {
+                    Included(x) => {
+                        assert!(self.secondary_index.remove(&(x, true)).unwrap().is_none())
+                    }
+                    Excluded(x) => {
+                        assert!(self.secondary_index.remove(&(x, false)).unwrap().is_none())
+                    }
+                    Unbounded => (),
+                }
+            }
+            while let Some((k, v)) = iter.next() {
+                match translate_lower_bound(v.0.clone()) {
+                    Included(x) => {
+                        if from_bottom || iter.peek().is_some() {
+                            self.secondary_index.remove(&x).unwrap(); // TODO: assert the epochs are
+                                                                      // the same?
+                        } else {
+                            // From top and last element, we should simply convert it to `None`
+                            // instead of removing
+                            let old = self.secondary_index.get_mut(&x).unwrap();
+                            if let Some(e) = old {
+                                assert_eq!(*e, *k);
+                            }
+                            *old = None;
+                        }
+                    }
+                    Unbounded => assert_eq!(self.lower_unbounded_epoch.take().unwrap(), *k),
+                    _ => unreachable!(),
+                }
+            }
+            Some((range, new_existing_range))
+        } else {
+            None
+        }
+    }
+
+    /// `access` is called when a range is accessed during an epoch. `access` assumes that
+    /// the ranges provided form a continuous path as a function of the epochs.
+    pub(crate) fn access(&mut self, range: ScalarRange, epoch: u64) {
+        assert_ne!(epoch, 0);
```

Comment:
Add comment:
We have two objectives:
1. Update the secondary index with the new range, removing any overlapping bounds
2. From the main index: remove the overlapping ranges, and in addition, shorten the ranges that partially overlap with the range at the end points (and deal with boundary conditions where the range's endpoints match exactly).